packages/phone-number-privacy/common/src/utils/sequential-delay.ts

```diff
@@ -0,0 +1,80 @@
+import { SequentialDelayDomain, SequentialDelayStage } from '../interfaces'
+interface IndexedSequentialDelayStage extends SequentialDelayStage {
+  // The stage's index in the stage array
+  index: number
+  // The attempt number at which the stage begins
+  start: number
+}
+export interface SequentialDelayResult {
+  accepted: boolean
+  state?: SequentialDelayState
+}
+interface SequentialDelayState {
+  // Timestamp used for deciding when the next request will be accepted.
+  timer: number
+  // Number of queries that have been accepted for the SequentialDelayDomain instance.
+  counter: number
+}
+
+export const checkSequentialDelay = (
+  domain: SequentialDelayDomain,
+  attemptTime: number,
+  state?: SequentialDelayState
+): SequentialDelayResult => {
+  const counter = state?.counter ?? 0
+  const timer = state?.timer ?? 0
+  const stage = getIndexedStage(domain, counter)
+
+  if (!stage) {
+    return { accepted: false, state }
+  }
+
+  const resetTimer = stage.resetTimer ?? true
+  const delay = getDelay(stage, counter)
+  const notBefore = timer + delay
+
+  if (attemptTime < notBefore) {
+    return { accepted: false, state }
+  }
+
+  // Request is accepted. Update the state.
+  return {
+    accepted: true,
+    state: {
+      counter: counter + 1,
+      timer: resetTimer ? attemptTime : notBefore,
+    },
+  }
+}
+
+const getIndexedStage = (
+  domain: SequentialDelayDomain,
+  counter: number
+): IndexedSequentialDelayStage | null => {
```

Comment:
As a general practice I've seen it suggested that only one of `null` or `undefined` should be used at any time. I don't feel particularly strongly on this, so you can feel free to ignore this if you have a stronger opinioin.
```suggestion
): IndexedSequentialDelayStage | undefined => {
```